[bugfix] Parameters to ignore is not being fully respected

YLGH opened the pull request:
Summary: I was experimenting with a model that had sharded tensors as parameters, and I set them to be ignored. However, DDP was still checking their device + some other sanitization logic. If we're ignoring them, then we shouldn't need to check

Test Plan: unit tests

Differential Revision: D37726515



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

awgu commented on torch/nn/parallel/distributed.py:
```diff
@@ -539,7 +539,15 @@ def __init__(
-        if not any((p.requires_grad for p in module.parameters())):
+
+        if hasattr(module, "_ddp_params_and_buffers_to_ignore"):
+            self.parameters_to_ignore = module._ddp_params_and_buffers_to_ignore
```
Do we need to make this into a (public) attribute? Can we also make it into a `set` given the membership checks for constructing `module_parameters`?
```suggestion
            parameters_to_ignore = set(module._ddp_params_and_buffers_to_ignore)
```

awgu commented on torch/nn/parallel/distributed.py:
```diff
@@ -539,7 +539,15 @@ def __init__(
+        else:
+            self.parameters_to_ignore = []
+
+        self.module_parameters = [param for name, param in module.named_parameters() if name not in self.parameters_to_ignore]
```
We should probably make this attribute private.

awgu commented on torch/nn/parallel/distributed.py:
```diff
@@ -600,16 +608,13 @@ def __init__(
         self.dim = dim
         self.module = module
-        self.device = list(self.module.parameters())[0].device
+        self.device = list(self.module_parameters)[0].device
```
No need for `list` when `self.module_parameters` is already a `list`.

Also, what happens if you ignore all of the module's parameters? We should handle that case gracefully instead of `self.device = ...` causing an index out of bounds error.

YLGH replied:
Added a

```
        if not self._module_parameters:
            self._log_and_throw(
                RuntimeError,
                "DistributedDataParallel is not needed when a module doesn't have"
                "any parameters that are not ignored."
            )

```
at the beginning,

awgu commented:
I will let others review/approve the PR since they are more familiar with DDP.

wanchaol commented on torch/nn/parallel/distributed.py:
```diff
@@ -539,7 +539,15 @@ def __init__(
-        if not any((p.requires_grad for p in module.parameters())):
+
+        if hasattr(module, "_ddp_params_and_buffers_to_ignore"):
+            self.parameters_to_ignore = module._ddp_params_and_buffers_to_ignore
```
nit: maybe call it `_params_buffers_to_ignore` to align what this actually mean?

YLGH replied:
@wanchaol there are a couple of internal usages that seem to check against self.parameters_to_ignore (or explicitly set this), maybe we can leave this part the same to not break BC?

wanchaol reviewed (commented):
looks good to me. Is it possible to add some proper testing on this change?

rohan-varma reviewed (commented):
LGTM, can we add some unittests, possibly reproducing the issue that we saw with shardedtensors?

YLGH commented:
Thanks folks, made changes based on suggestions ~ 

I'm having a bit of trouble testing, will ping y'all,

one quick question I have is why do we need to set _ddp_params_and_buffers_to_ignore as a DistributeDataParallel static method? Why can't we just have DistributeDataParallel.__init__ take this in as a kwarg? (params_and_buffers to ignore)?

YLGH commented:
> 

@rohan-varma , there are some existing tests covering _set_params_and_buffers_to_ignore_for_model that are green. Not sure how straightforward it would be to add a test with ShardedTensor since (really) it's an issue w/ DDP and ST not composing super well together. *I could add a test for that, but then we'd have some unnecessary coupling from DistributedDataParallel unit tests w/ ShardedTensor

YLGH committed:
[bugfix] Parameters to ignore is not being fully respected (#81136)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/81136

I was experimenting with a model that had sharded tensors as parameters, and I set them to be ignored. However, DDP was still checking their device + some other sanitization logic. If we're ignoring them, then we shouldn't need to check

Test Plan: unit tests

Differential Revision: D37726515

fbshipit-source-id: 2983d728cfd4e174817fd732d07ae22addde9c7f

wanchaol commented:
> > 
> 
> @rohan-varma , there are some existing tests covering _set_params_and_buffers_to_ignore_for_model that are green. Not sure how straightforward it would be to add a test with ShardedTensor since (really) it's an issue w/ DDP and ST not composing super well together. *I could add a test for that, but then we'd have some unnecessary coupling from DistributedDataParallel unit tests w/ ShardedTensor

@YLGH Since it's composability problem with ShardedTensor, maybe we can add a unit test in this file, similar with `test_with_ddp`, and see if the parameters are being ignored correctly? https://github.com/pytorch/pytorch/blob/4124307faeed923175d3e02fa685c0788e99f0ee/test/distributed/_shard/test_replicated_tensor.py#L206

Or if it's pretty hard to add tests checking this, @rohan-varma are we confident that the CI passed then it should be good for us to land this PR?

rohan-varma reviewed (commented):
This LGTM overall, but I'd prefer adding some tests:

1. Regular nn.Module where ignored params might be on different device or something that might've thrown before. 
2. Test with ShardedTensors.
